### ISSUE TITLE
Roslyn fails to detect local functions that violate their nullability contracts specified via System.Diagnostics.CodeAnalysis attributes

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#55396]
**Version Used**: 
```
Branch main (2 Aug 2021)
Latest commit 914d69c by Julien Couvreur:
Track "raw string literals" feature (#55307)
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
using System;
using System.Diagnostics.CodeAnalysis;

string? str = null;
Fail();
Console.WriteLine(str.Length);

[DoesNotReturn] static void Fail() { }
```

**Expected Behavior**:
`CS8763: A method marked [DoesNotReturn] should not return.` is reported for the local function `Fail`

**Actual Behavior**:
No warnings at all. The code crashes at runtime with a `NullReferenceException`

**Notes**:
If you rewrite the local function as a static method Roslyn correctly reports the issue.
The problem seems to affect most annotations on local functions, here are a few more examples:

With `[NotNullWhen]`:
```cs
#nullable enable
using System;
using System.Diagnostics.CodeAnalysis;

if (TryGetValue(out var value))
{
  Console.WriteLine(value.Length);
}

static bool TryGetValue([NotNullWhen(true)] out string? value)
{
  value = null;
  return true;
}
```
Expected: `CS8762: Parameter 'value' must have a non-null value when exiting with 'true'.` for the local function's return value
Actual: no warnings. The code crashes at runtime with a `NullReferenceException`

With `[NotNull]`:
```cs
#nullable enable
using System;
using System.Diagnostics.CodeAnalysis;

string? str = null;
AssertNotNull(str);
Console.WriteLine(str.Length);

static void AssertNotNull([NotNull] string? value)
{
}
```
Expected: `CCS8777: Parameter 'value' must have a non-null value when exiting.` for the local function's exit point
Actual: no warnings. The code crashes at runtime with a `NullReferenceException`

**Comment by RikkiGibson** [comment]
Duplicate of #47896
